File: medium/279. Perfect Squares/279. Perfect Squares.ts
function numSquares(n: number): number {
    const dp = new Array<number>(n + 1).fill(n)
    dp[0] = 0

    for (let target = 1; target <= n; target++) {
        for (let i = 1; i <= target; i++) {
            const square = i ** 2
            if (square > target) break
            dp[target] = Math.min(dp[target], (dp[target - square] || 0) + 1)
        }
    }

    return dp[n]
};

File: medium/279. Perfect Squares/279. Perfect Squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [n] * (n + 1)
        dp[0] = 0

        for i in range(1, n + 1):
            for j in range(1, i + 1):
                square = j**2
                if square > n:
                    break

                dp[i] = min(dp[i], dp[i - square] + 1)

        return dp[n]


solution = Solution()
print(solution.numSquares(n=12))
print(solution.numSquares(n=13))
